hazelcast/src/main/java/com/hazelcast/sql/impl/SqlServiceImpl.java

```diff
@@ -174,10 +174,16 @@ private SqlCursor query0(String sql, List<Object> params, long timeout, int page
             }
 
             Plan plan = prepare(unwrappedSql);
+            if (plan == null) {
+                return new SingleValueSqlCursor(0);
+            }
 
             state = internalService.executeExplain(plan);
         } else {
             Plan plan = prepare(sql);
+            if (plan == null) {
```

Comment:
Here we make the purpose of optimization stage fuzzy, because sometimes optimizer does optimization, sometimes it does parsing and execution. I would prefer to retain clear boundaries for component responsibilities so that the optimizer prepares the plan, and then in the `SqlServiceImpl` we decide how to execute the returned plan. This is needed to ensure that we can evolve the engine easily. For example, consider that in future we will need to implement SQL query statistics or asynchronous query cancellation for JDBC. With a centralized approach, it will be much easier than trying to hook into disparate execution points. 

I do not want to slow down the prototyping due to this comment, so I propose to do the following: 
1) Go on with the current approach for now
2) Soon I will switch to API implementation, where I can try to decouple execution from planning in the following way:
- Planner return a plan that could be SELECT plan, DDL plan, etc
- Then depending on the plan type, we feed it to a designated handler in the "core" module